File: ios/Classes/MlKit/InitialScannerParameters.swift
//
//  InitialScannerParameters.swift
//  mlkit_scanner
//
//  Created by ООО "ДНС Технологии" on 13.04.2023.
//

import AVFoundation
import Foundation

/// Initial scanner settings.
///
/// Will be applied to the scanner during initialization.
struct InitialScannerParameters {
    /// Camera initial zoom.
    let initialZoom: Double?
    /// Scanner initial scanning area.
    let initialCropRect: CropRect?
    /// Initial camera.
    let initialCamera: CameraData?
    
    init?(arguments: Dictionary<String, Any?>?) {
        if (arguments == nil) {
            return nil
        }
        
        guard let initialZoom = arguments!["initialZoom"] as? Double?,
              let initialCropRect = arguments!["initialCropRect"] as? Dictionary<String, CGFloat>?,
              let initialCamera = arguments!["initialCamera"] as? Dictionary<String, Any?>? else {
            return nil
        }
        self.initialZoom = initialZoom
        self.initialCropRect = initialCropRect != nil ? CropRect(arguments: initialCropRect!): nil
        self.initialCamera = initialCamera != nil ? CameraData(arguments: initialCamera!) : nil
    }
}

File: android/src/main/kotlin/com/dns_technologies/mlkit_scanner/scanner/models/InitialScannerParameters.kt
package com.dns_technologies.mlkit_scanner.scanner.models

/**
 * Describes initial parameters used during scanner initialization.
 *
 * @property zoom Optional normalized camera zoom value.
 * @property cropRect Optional initial scanner recognition area.
 */
data class InitialScannerParameters(
    val zoom: Double? = null,
    val cropRect: RecognizeVisorCropRect? = null,
) {
    /** Creates [InitialScannerParameters] from a method channel argument map. */
    constructor(map: Map<String, Any?>) : this(
        (map["initialZoom"] as Double?),
        (map["initialCropRect"] as? Map<*, *>)?.let(RecognizeVisorCropRect::fromMap),
    )
}
